libcnb-cargo/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+    }?;
+
+    let mut cargo_args = vec!["build", "--target", target_triple.as_ref()];
+    match cargo_profile {
+        CargoProfile::Dev => {}
+        CargoProfile::Release => cargo_args.push("--release"),
+    }
+
+    let exit_status = Command::new("cargo")
+        .args(cargo_args)
+        .envs(
+            cross_compile::cross_compile_env(target_triple.as_ref())
+                .map_err(BuildError::CrossCompileError)?,
+        )
+        .current_dir(&project_path)
+        .spawn()
+        .and_then(|mut child| child.wait())
+        .map_err(BuildError::IoError)?;
+
+    if exit_status.success() {
+        let binary_path = cargo_metadata
+            .target_directory
+            .join(target_triple.as_ref())
+            .join(match cargo_profile {
+                CargoProfile::Dev => "debug",
+                CargoProfile::Release => "release",
+            })
+            .join(&target.name)
+            .into_std_path_buf();
+
+        Ok(binary_path)
+    } else {
+        Err(BuildError::UnexpectedExitStatus(exit_status))
+    }
+}
+
+#[derive(Debug)]
+pub enum BuildError {
+    IoError(std::io::Error),
+    UnexpectedExitStatus(ExitStatus),
+    CrossCompileError(CrossCompileError),
+    NoTargetsFound,
+    MultipleTargetsFound,
+    MetadataError(cargo_metadata::Error),
+    CouldNotFindRootPackage,
+}
+
+#[derive(Copy, Clone)]
+pub enum CargoProfile {
+    Dev,
+    Release,
+}
+
+/// Reads buildpack data from the given project path.
+///
+/// # Errors
+/// Will return `Err` if the buildpack data could not be read successfully.
+pub fn read_buildpack_data(
+    project_path: impl AsRef<Path>,
+) -> Result<BuildpackData<Option<toml::Value>>, BuildpackDataError> {
+    let buildpack_toml_path = project_path.as_ref().join("buildpack.toml");
+
+    fs::read_to_string(&buildpack_toml_path)
+        .map_err(BuildpackDataError::IoError)
+        .and_then(|file_contents| {
+            toml::from_str(&file_contents).map_err(BuildpackDataError::DeserializationError)
+        })
+        .map(|buildpack_toml| BuildpackData {
+            buildpack_toml_path,
+            buildpack_toml,
+        })
+}
+
+#[derive(Debug)]
+pub enum BuildpackDataError {
+    IoError(std::io::Error),
+    DeserializationError(toml::de::Error),
+}
+
+pub struct BuildpackData<BM> {
+    pub buildpack_toml_path: PathBuf,
+    pub buildpack_toml: BuildpackToml<BM>,
+}
+
+/// Creates a buildpack directory and copies all buildpack assets to it.
+///
+/// Assembly of the directory follows the constraints set by the libcnb framework. For example,
+/// the buildpack binary is only copied once and symlinks are used to refer to it when the CNB
+/// spec requires different file(name)s.
+///
+/// This function will not validate if the buildpack descriptor at the given path is valid and will
+/// use it as-is.
+///
+/// # Errors
+/// Will return `Err` if the buildpack directory could not be assembled.
+pub fn assemble_buildpack_directory(
```

Comment:
In the voice of Captain America: Buildpack directory ASSEMBLE!

Reply:
![image](https://user-images.githubusercontent.com/177877/144211096-9e896c59-c8e9-4476-9f62-6af59c4eaf57.png)